rita/src/rita_common/tunnel_manager/mod.rs

```diff
@@ -59,48 +101,80 @@ pub struct Tunnel {
     pub listen_ifidx: u32,      // the physical interface this tunnel is listening on
     pub listen_port: u16,       // the local port this tunnel is listening on
     pub localid: LocalIdentity, // the identity of the counterparty tunnel
+    state: TunnelState,
 }
 
 impl Tunnel {
     fn new(
         ip: IpAddr,
+        iface_name: String,
         our_listen_port: u16,
         ifidx: u32,
         their_id: LocalIdentity,
-    ) -> Result<Tunnel, Error> {
-        let iface_name = KI.setup_wg_if().unwrap();
-
-        let tunnel = Tunnel {
+    ) -> Tunnel {
```

Comment:
ok I'm concerned about this, we need to run new -> open -> monitor and in tests you just run new -> monitor that's the idea right? 

My big concern is that I designed new to encapsulate everything so that it was impossible for us to have a tunnel object without having the real tunnel. This reduces the risk of desyncing of the real system tunnel state from what we have represented here. I guess we only actually create tunnels in one spot so it shouldn't be too hard to keep straight. But that's my worry.

Reply:
There is no safety measurements between opening a tunnel and inserting a tunnel in the hashmap. After opening a tunnel there are few steps that might fail - `TcpStream::connect`, `tunnel.restrict`, `tunnel.monitor`. What if we'd add new state i.e. "TunnelState::New" before openning a tunnel, and change it to "TunnelState::NotRegistered" after all steps will pass? Of course `TunnelState::New` would be already in the internal hashmap. This way you could wipe tunnels which are in `TunnelState::New` for too long? Or just simply wipe the tunnel in any case one of the operations would fail?

Reply:
why don't we wait to insert the tunnel until these operations have succeeded? That way we don't need any garbage collection on the tunnels list.

Reply:
@jkilpatr my point is that first we do `open_tunnel` which already does changes in the system. If any further operation before inserting into hashmap fails we can have a tunnel which doesn't reflect our state, becase we return early with Err (thanks to `?` operator). What do you think about such case?